File: apps/java/design-pattern/src/main/java/pattern/singleton/SimpleLogger.java
package pattern.singleton;

/**
 * シンプルなログクラス。
 * テスト容易性のために StringBuilder に出力する。
 */
public class SimpleLogger {
    public static final int ERROR = 1;
    public static final int WARNING = 2;
    public static final int INFO = 3;

    private final StringBuilder output = new StringBuilder();
    private int level = INFO;

    public void setLevel(int level) {
        this.level = level;
    }

    public int getLevel() {
        return level;
    }

    public void error(String msg) {
        if (level >= ERROR) {
            output.append("[ERROR] ").append(msg).append("\n");
        }
    }

    public void warning(String msg) {
        if (level >= WARNING) {
            output.append("[WARNING] ").append(msg).append("\n");
        }
    }

    public void info(String msg) {
        if (level >= INFO) {
            output.append("[INFO] ").append(msg).append("\n");
        }
    }

    public String getLoggedContent() {
        return output.toString();
    }
}

File: apps/ruby/design-pattern/13-singleton/lib/simple_logger.rb
require "stringio"

class SimpleLogger
  ERROR = 1
  WARNING = 2
  INFO = 3

  attr_accessor :level
  attr_reader :output

  def initialize
    @output = StringIO.new
    @level = INFO
  end

  def error(message)
    @output.puts("[ERROR] #{message}") if @level >= ERROR
  end

  def warning(message)
    @output.puts("[WARNING] #{message}") if @level >= WARNING
  end

  def info(message)
    @output.puts("[INFO] #{message}") if @level >= INFO
  end

  def logged_content
    @output.string
  end
end
